### PR TITLE
Early-stage Programmatic API / Custom VM Option

### PR DESCRIPTION
**Pull request by holgerd77**
Phew, this PR gives me some inner peace of mind 😄  since it should both satisfy the demand from @jochem-brouwer to keep the more complex instantiations on `Blockchain` or `VM` out of `Config` and at the same time shows that a custom VM integration is nevertheless pretty much doable, which was my greatest fear that we would loose on that.

It then goes further and adds a first early-stage programmatic API for the client. I find this highly useful to expose this early, since this should hopefully trigger experimentation from the community and encourage feedback on this, so that it is assured that we take appropriate design paths here and think this along from the beginning.

Some notes on the changes:

- Moved VM from `FullEthereumService` to `FullSynchronizer`
- Switched to local `./datadir` default datadir from defaulting to the [HOME]/Library/Ethereum directory (this is far too dangerous to accidentally overwriting a whole blockchain when invoking a client instance programtically or within the tests)
- Added VM option to Config
- Renamed `Node` to `EthereumClient` to make the entrypoint to the application more obvious (this was actually no quick shot but already in discussion with Vinay two years ago, think this is now becoming more and more obvious when we want to expose the client programatically, `Node` is generally not a lucky name in the `Node.js` world 🙂 )
- I have moved the index.ts export file to old_index.ts, for now I never stumbled upon a practical use case for the old-version exports, but might be worth to keep in case there is some hidden one. A new index.ts file now exposes the public API classes for an easy import

### CODE REVIEW

File: lib/sync/fullsync.ts [hunk 1]
```diff
@@ -10,11 +11,24 @@ import { BlockFetcher } from './fetcher'
...
+
+    if (!this.config.vm) {
+      this.vm = new VM({
+        common: this.config.common,
+        blockchain: this.chain.blockchain,
+      })
+    } else {
+      this.vm = this.config.vm
+      //@ts-ignore blockchain has readonly property
+      this.vm.blockchain = this.chain.blockchain
```
Comment (jochem-brouwer) [inline lib/sync/fullsync.ts:30]: CodeCov reports that this is not covered by tests, could you add a test that ensures that the configured VM is indeed the correct one?

**Review by jochem-brouwer — COMMENTED** [review]
Looks good to me, except that we want to test that the configured VM is indeed the correct one. This was much more straightforward than I thought, I assumed it would lead to many problems 😅

**Comment by holgerd77** [comment]
I've now also added a custom VM test case 2d5ceee55fb2ace11487bd8dcc2ef5c234ff8662, this is now ready for re-review! 😄

**Review by jochem-brouwer — APPROVED** [review]
Great!

**Comment by ryanio** [comment]
looks great, thanks for detailing the changes. I will update the diagram in a new PR.

**Pull request merged**